tensorflow_datasets/image/cityscapes.py

```diff
@@ -0,0 +1,280 @@
+
+  Ingored examples:
+  - For 'cityscapes/stereo_disparity_extra':
+    - troisdorf_000000_000073_{*} images (no disparity map present)
+
+  WARNING: this dataset requires users to setup a login and password in order to get the files.
+'''
+
+# TODO add instance ids (might need to import cityScapesScripts)
+
+class CityscapesConfig(tfds.core.BuilderConfig):
+  '''BuilderConfig for Cityscapes
+
+    Args:
+      right_images (bool): Enables right images for stereo image tasks.
+      segmentation_labels (bool): Enables image segmentation labels.
+      disparity_maps (bool): Enables disparity maps.
+      train_extra_split (bool): Enables train_extra split. This automatically
+          enables coarse grain segmentations, if segmentation labels are used.
+  '''
+
+  @api_utils.disallow_positional_args
+  def __init__(self, right_images=False, segmentation_labels=True,
+               disparity_maps=False, train_extra_split=False, **kwargs):
+    super().__init__(**kwargs)
+
+    self.right_images = right_images
+    self.segmentation_labels = segmentation_labels
+    self.disparity_maps = disparity_maps
+    self.train_extra_split = train_extra_split
+
+    self.ignored_ids = set()
+
+    # Setup required zips and their root dir names
+    self.zip_root = {}
+    self.zip_root['images_left'] =\
+      ('leftImg8bit_trainvaltest.zip', 'leftImg8bit')
+
+    if self.train_extra_split:
+      self.zip_root['images_left/extra'] =\
+        ('leftImg8bit_trainextra.zip', 'leftImg8bit')
+
+    if self.right_images:
+      self.zip_root['images_right'] =\
+        ('rightImg8bit_trainvaltest.zip', 'rightImg8bit')
+      if self.train_extra_split:
+        self.zip_root['images_right/extra'] =\
+          ('rightImg8bit_trainextra.zip', 'rightImg8bit')
+
+    if self.segmentation_labels:
+      if not self.train_extra_split:
+        self.zip_root['segmentation_labels'] =\
+          ('gtFine_trainvaltest.zip', 'gtFine')
+        self.label_suffix = 'gtFine_labelIds'
+      else:
+        # The 'train extra' split only has coarse labels unlike train and val.
+        # Therefore, for consistency across splits, we also enable coarse labels
+        # using the train_extra_split flag.
+        self.zip_root['segmentation_labels'] = ('gtCoarse.zip', 'gtCoarse')
+        self.zip_root['segmentation_labels/extra'] = \
+          ('gtCoarse.zip', 'gtCoarse')
+        self.label_suffix = 'gtCoarse_labelIds'
+
+    if self.disparity_maps:
+      self.zip_root['disparity_maps'] =\
+        ('disparity_trainvaltest.zip', 'disparity')
+      if self.train_extra_split:
+        self.zip_root['disparity_maps/extra'] =\
+          ('disparity_trainextra.zip', 'disparity')
+        self.ignored_ids.add('troisdorf_000000_000073') # No disparity for this file
+
+
+class Cityscapes(tfds.core.GeneratorBasedBuilder):
+  '''Base class for Cityscapes datasets'''
+
+  BUILDER_CONFIGS = [
+      CityscapesConfig(
+          name='semantic_segmentation',
```

Comment:
It was a deliberate choice - Disparity maps need additional permissions from the authors and aren't readily available. 'Extra' splits - whether for segmentation or for the disparity requires additional files (50GB). I chose this route so that each configuration so that it would work for all cases.

Reply:
I gave this some more thought today to see if I could refactor the code to use fewer if/else statements. The problem arises because Cityscapes has too many features that every user may not use.

1. There's two 'datasets' for different tasks - semantic semgentation and disparity inference. Disparity tasks can also require the right images - another additional download.
2. Every cityscapes 'dataset' (semantic seg / stereo depth) has an additional 'extra' split. This split is much larger and any labels provided are coarse grain instead.

I think we still need 4 (2 (task) x 2 (extra split)) configurations.

One feature that might reduce some if/else might be to use a constant feature dict and use a 'default' image tensor (since cityscape labels are images) if this is acceptable.

Reply:
Are there any tutorials on how to use this class please?

Reply:
You just have to do something along the lines of:

```
import tensorflow_datasets as tfds
dataset = tfds.load(name="cityscapes/semantic_segmentation_extra", split="train")
# Can also specify config from above.
```

I'd give a working Colab but unfortunately this is a pretty big dataset and requires manual download :)

Reply:
@f1recracker thank you for the reply! 
I downloaded the dataset manually and will attempt to access it with the syntax you suggested
Thanks again.